relayer/crates/starknet-chain-components/src/types/cosmos/client_state.rs

```diff
@@ -173,3 +181,33 @@ impl From<CometClientState> for Any {
         .into()
     }
 }
+
+pub struct EncodeChainId;
+
+impl<Encoding, Strategy> MutEncoder<Encoding, Strategy, ChainId> for EncodeChainId
+where
+    Encoding: CanEncodeMut<Strategy, String>,
+{
+    fn encode_mut(
+        encoding: &Encoding,
+        chain_id: &ChainId,
+        buffer: &mut Encoding::EncodeBuffer,
+    ) -> Result<(), Encoding::Error> {
+        let chain_id_str = chain_id.as_str().to_string();
+        encoding.encode_mut(&chain_id_str, buffer)?;
+        Ok(())
+    }
+}
+
+impl<Encoding, Strategy> MutDecoder<Encoding, Strategy, ChainId> for EncodeChainId
+where
+    Encoding: CanDecodeMut<Strategy, String> + CanRaiseError<&'static str>,
+{
+    fn decode_mut<'a>(
+        encoding: &Encoding,
+        buffer: &mut Encoding::DecodeBuffer<'a>,
+    ) -> Result<ChainId, Encoding::Error> {
+        let chain_id_str = encoding.decode_mut(buffer)?;
+        ChainId::new(&chain_id_str).map_err(|_| Encoding::raise_error("invalid chain id"))
+    }
+}
```

Comment:
Is this correct to satisfy `CanEncodeAndDecodeMut<ViaProtobuf, ChainId>`?

Reply:
I have explained this to @ljoss17. This encoding strategy works for Cairo, but not for protobuf. This is because protobuf requires special field tag to be applied to field values, and "primitive" values like strings can only be field values, meaning they always require a tag.

The solution is to define a `EncodeChainIdField<const TAG: u32>` encoder, which makes use of `EncodeStringField<TAG>` to encode the given protobuf field.